lib/sqlalchemy/dialects/postgresql/asyncpg.py

```diff
@@ -644,11 +644,12 @@ class AsyncAdapt_asyncpg_connection(AdaptedConnection):
         "_prepared_statement_cache",
         "_invalidate_schema_cache_asof",
         "_execute_mutex",
+        "_name_func"
     )
 
     await_ = staticmethod(await_only)
 
-    def __init__(self, dbapi, connection, prepared_statement_cache_size=100):
+    def __init__(self, dbapi, connection, prepared_statement_cache_size=100, name_func=None):
```

Comment:
I think name_func is too generic, expecially on the engine.

Let's call this `pepared_statement_name_func` or similar (@zzzeek suggestions here?)

Reply:
yeah for sure, my naming choice is quite wide

Reply:
I would use prepared_statement_name_func.

here is the problem.   this scheme will create endless numbers of prepared statements on pooled connections that are not reused.   I think if one is using this scheme, it's very important that [discard](https://www.postgresql.org/docs/current/sql-discard.html) is used appropriately to at the very least release prepared statements ( as well as transactional state as always)   I guess if pgbouncer is used, then on the client (here) we are using NullPool and pgbouncer will have DISCARD set up appropriately but we would to make sure this is mentioned (1. use nullpool w/ pgbouncer; 2. make sure DISCARD is set up).

@teror4uks is that the configuration you are using?    basically with this param I want to have an understanding of a working context for it.

Reply:
yes you are right we are using NullPool and pgbouncer, the common setup looks like this 

```python
engine = create_async_engine(
    url,
    poolclass=NullPool,
    connect_args={
        'statement_cache_size': 0,
        'connection_class': CConnection,
    },
)
```

Reply:
OK so when we document this, we should put that example in place and definitely mention that these prepared statements will build up like lightning if they are not cleared out